### PR TITLE
Fix vote

### LINKED ISSUE #1790: Vote without Register the candidate, then the Account is not able to Transfer NEO
**Describe the bug**
When the account vote for an account which is not candidate yet, then relay the vote transaction, then try to send some neo to other accounts. it fails.

**To Reproduce**
1. Open a wallet, send 100 neo(to vote)/100 gas (ensure it has enough fee to pay)
58f2d080a469c1e08075280cd3c7d4b

2. Then vote, choose the wallet default address and default pubkey
3. Then relay the transaction.
6bccfcb396e1505783cc7b2db565332

4. send neo to other account(not the default address)
bef33d1569da4fa43072971feeec808

**Expected behavior**
send neo successful.

**Platform:**
 - OS: [Windows 10 x64
 - Version master 

Root Cause Analysis:

https://github.com/neo-project/neo/blob/5406085af441bdf916a735b676bfbc67cf10cdc9/src/neo/SmartContract/Native/Tokens/NeoToken.cs#L40-L49

because CreateStorageKey(Prefix_Candidate).Add(state.VoteTo) is null.

debug version: 
```cs
        protected override void OnBalanceChanging(ApplicationEngine engine, UInt160 account, NeoAccountState state, BigInteger amount)
        {
            Console.WriteLine(account.ToString());
            Console.WriteLine(state.ToString());
            DistributeGas(engine, account, state);
            if (amount.IsZero) return;
            if (state.VoteTo != null)
            {
                Console.WriteLine(state.VoteTo.ToString());
                if (CreateStorageKey(Prefix_Candidate) is null)
                    Console.WriteLine("create storage key is null");
                if (CreateStorageKey(Prefix_Candidate).Add(state.VoteTo) is null)
                    Console.WriteLine("add state.voto is null");
                if (engine.Snapshot.Storages.GetAndChange(CreateStorageKey(Prefix_Candidate).Add(state.VoteTo)) is null)
                    Console.WriteLine("can not find state.voteTo Storage");
                engine.Snapshot.Storages.GetAndChange(CreateStorageKey(Prefix_Candidate).Add(state.VoteTo)).GetInteroperable<CandidateState>().Votes += amount;
                engine.Snapshot.Storages.GetAndChange(CreateStorageKey(Prefix_VotersCount)).Add(amount);
            }
        }
```

https://github.com/neo-project/neo/blob/5406085af441bdf916a735b676bfbc67cf10cdc9/src/neo/SmartContract/Native/Tokens/NeoToken.cs#L165-L172

### PR DESCRIPTION
**Pull request by Tommo-L**
Close #1790

### CODE REVIEW

**Comment by cloud8little** [comment]
Test passed:
1. send wallet a.json with 100 neo/100 gas.
2. vote for himself.
3. send neo to another account successfully.
4. register for himself to be a candidate.
5. vote successful and get candidates.

82a98bf3a77ed98bbdd4dffb6984222
9a1b44c86a95ccce5dd04d59827c024

File: src/neo/SmartContract/Native/Tokens/NeoToken.cs [hunk 1]
```diff
@@ -172,6 +171,7 @@ private bool Vote(ApplicationEngine engine, UInt160 account, ECPoint voteTo)
                 if (!state_validator.Registered) return false;
                 state_validator.Votes += state_account.Balance;
             }
+            state_account.VoteTo = voteTo;
```
Comment (shargon) [inline src/neo/SmartContract/Native/Tokens/NeoToken.cs:174]: Could you add a UT for ensure this case? @Tommo-L

Comment (Tommo-L) [reply src/neo/SmartContract/Native/Tokens/NeoToken.cs:174]: Okay, I'll add it.

**Comment by superboyiii** [comment]
@shargon Ready for merge?

**Comment by cloud8little** [comment]
Retesting.

**Comment by cloud8little** [comment]
Test passed:
1. Register a candidate not exist in the wallet.
```
NEO-CLI v3.0.0-preview2  -  NEO v3.0.0-preview2  -  NEO-VM v3.0.0-preview2

neo> open wallet wallet_1.json
password: ********
neo> list asset
Nc56Zrq72VN9Xr6wo3HyFn3oFZUvoYQJtj
NEO: 0
GAS: 0

Ncu47NyTJLJPEFi859gmuqQ8fPPt1AVwHX
NEO: 100000000
GAS: 30000000

----------------------------------------------------
Total:   NEO: 100000000    GAS: 30000000

NEO hash: 0xde5f57d430d3dece511cf975a8d37848cb9e0525
GAS hash: 0x668e0c1f9d7b70a99dd9e06eadd4c784d641afbc
neo> register candidate Ncu47NyTJLJPEFi859gmuqQ8fPPt1AVwHX
Error: Missing the signature of 026a59b03f981a5dfded185472f4f19409cd229b5cfd11721e70f2b1d16079376a
```

2. Register candidate

```
neo> list key
  Address: Nc56Zrq72VN9Xr6wo3HyFn3oFZUvoYQJtj
PublicKey: 026a59b03f981a5dfded185472f4f19409cd229b5cfd11721e70f2b1d16079376a

  Address: Nc56Zrq72VN9Xr6wo3HyFn3oFZUvoYQJtj
PublicKey: 026a59b03f981a5dfded185472f4f19409cd229b5cfd11721e70f2b1d16079376a

neo> send neo NUzy2y33pa7DvUgnihxdqSPUTZi3Hb3DnB 100
password: ********
TXID: 0x2c045472e8de39fe3d36101f2f98b0d2c94d6c47a4546d03bd2372243b2b6a3a
neo> send gas NUzy2y33pa7DvUgnihxdqSPUTZi3Hb3DnB 100
password: ********
TXID: 0xdca11fa586665dd4dd2bdb403414a59bee8bb5b1a8ea89f7b68212d3fd6f50e0
neo> open wallet a.json
password: *
neo> list asset
NUzy2y33pa7DvUgnihxdqSPUTZi3Hb3DnB
NEO: 100
GAS: 100

----------------------------------------------------
Total:   NEO: 100    GAS: 100

NEO hash: 0xde5f57d430d3dece511cf975a8d37848cb9e0525
GAS hash: 0x668e0c1f9d7b70a99dd9e06eadd4c784d641afbc
neo> list key
  Address: NUzy2y33pa7DvUgnihxdqSPUTZi3Hb3DnB
PublicKey: 02ae840030e1e58d3c8b08568e13075372fb17c111fb5690fc3c26a7342abc2a03

neo> register candidate NUzy2y33pa7DvUgnihxdqSPUTZi3Hb3DnB
Invoking script with: '0c2102ae840030e1e58d3c8b08568e13075372fb17c111fb5690fc3c26a7342abc2a0311c00c11726567697374657243616e6469646174650c1425059ecb4878d3a875f91c51ceded330d4575fde41627d5b52'
VM State: HALT
Gas Consumed: 0.0600775
Evaluation Stack: [{"type":"Any"}]

relay tx(no|yes): no
neo> register candidate NUzy2y33pa7DvUgnihxdqSPUTZi3Hb3DnB
Invoking script with: '0c2102ae840030e1e58d3c8b08568e13075372fb17c111fb5690fc3c26a7342abc2a0311c00c11726567697374657243616e6469646174650c1425059ecb4878d3a875f91c51ceded330d4575fde41627d5b52'
VM State: HALT
Gas Consumed: 0.0600775
Evaluation Stack: [{"type":"Any"}]

relay tx(no|yes): yes
Signed and relayed transaction with hash=0x02d47d1558cd78280ddd7120558b69dd02900122a61443944d27926e5a46a470
neo> get candidates
Invoking script with: '10c00c0d67657443616e646964617465730c1425059ecb4878d3a875f91c51ceded330d4575fde41627d5b52'
VM State: HALT
Gas Consumed: 1.0100757

Candidates:
02ae840030e1e58d3c8b08568e13075372fb17c111fb5690fc3c26a7342abc2a03      0
neo> list key
  Address: NUzy2y33pa7DvUgnihxdqSPUTZi3Hb3DnB
PublicKey: 02ae840030e1e58d3c8b08568e13075372fb17c111fb5690fc3c26a7342abc2a03
```
3. Unregister candidate and get candidates

```
neo> unregister candidate Nc56Zrq72VN9Xr6wo3HyFn3oFZUvoYQJtj
error: The address 0x7646c5c571b77ab316815d05d1c2ede2e9ea3bb1 was not found in the wallet
neo> unregister candidate NUzy2y33pa7DvUgnihxdqSPUTZi3Hb3DnB
Invoking script with: '0c2102ae840030e1e58d3c8b08568e13075372fb17c111fb5690fc3c26a7342abc2a0311c00c13756e726567697374657243616e6469646174650c1425059ecb4878d3a875f91c51ceded330d4575fde41627d5b52'
VM State: HALT
Gas Consumed: 0.0600775
Evaluation Stack: [{"type":"Any"}]

relay tx(no|yes): yes
Signed and relayed transaction with hash=0x0ae9442300662e26530ae0f6051d4181747f12f68dbd7345e972f11e22028875
neo> get candidate
error: Command not found
neo> get candidates
Invoking script with: '10c00c0d67657443616e646964617465730c1425059ecb4878d3a875f91c51ceded330d4575fde41627d5b52'
VM State: HALT
Gas Consumed: 1.0100757
```
4. Unregister a non-candidate

```
neo> unregister candidate NUzy2y33pa7DvUgnihxdqSPUTZi3Hb3DnB
Error: The candidate does not exist
```
5. Vote a non-candidate and send neo to other account.
```
neo> vote NUzy2y33pa7DvUgnihxdqSPUTZi3Hb3DnB 02ae840030e1e58d3c8b08568e13075372fb17c111fb5690fc3c26a7342abc2a03
Error: The candidate does not exist
neo> send neo Nc56Zrq72VN9Xr6wo3HyFn3oFZUvoYQJtj 1
password: *
TXID: 0x0606e732e534474d1a3949dde42304612b1cf8081cfe923957b783b41c2cdc94
neo> list asset
NUzy2y33pa7DvUgnihxdqSPUTZi3Hb3DnB
NEO: 99
GAS: 99.7525614

----------------------------------------------------
Total:   NEO: 99    GAS: 99.7525614

NEO hash: 0xde5f57d430d3dece511cf975a8d37848cb9e0525
GAS hash: 0x668e0c1f9d7b70a99dd9e06eadd4c784d641afbc
```
5. Then register candidate and vote to the candidate; get candidates and check the votes.

```
neo> register candidate NUzy2y33pa7DvUgnihxdqSPUTZi3Hb3DnB
Invoking script with: '0c2102ae840030e1e58d3c8b08568e13075372fb17c111fb5690fc3c26a7342abc2a0311c00c11726567697374657243616e6469646174650c1425059ecb4878d3a875f91c51ceded330d4575fde41627d5b52'
VM State: HALT
Gas Consumed: 0.0600775
Evaluation Stack: [{"type":"Any"}]

relay tx(no|yes): yes
Signed and relayed transaction with hash=0x9deab808316d09794e1f33f0c25d142787b0f850aebb52403b52005fabe55fd5

neo> vote NUzy2y33pa7DvUgnihxdqSPUTZi3Hb3DnB 02ae840030e1e58d3c8b08568e13075372fb17c111fb5690fc3c26a7342abc2a03
Invoking script with: '0c2102ae840030e1e58d3c8b08568e13075372fb17c111fb5690fc3c26a7342abc2a030c1463ab15dab11bfa8ebc0b3ca0e2bec6300b45e35d12c00c04766f74650c1425059ecb4878d3a875f91c51ceded330d4575fde41627d5b52'
VM State: HALT
Gas Consumed: 5.0100793
Evaluation Stack: [{"type":"Any"}]

relay tx(no|yes): yes
Signed and relayed transaction with hash=0x5e737b6b0d9864823abaccd57812b6106a7ad3cb6773fb4e001d7b7d1659d081
neo> get candidates
Invoking script with: '10c00c0d67657443616e646964617465730c1425059ecb4878d3a875f91c51ceded330d4575fde41627d5b52'
VM State: HALT
Gas Consumed: 1.0100757

Candidates:
02ae840030e1e58d3c8b08568e13075372fb17c111fb5690fc3c26a7342abc2a03      99
```

**Comment by superboyiii** [comment]
@erikzhang Could you review it again?

File: src/neo/SmartContract/Native/Tokens/NeoToken.cs [hunk 2]
```diff
@@ -141,10 +141,15 @@ private bool UnregisterCandidate(ApplicationEngine engine, ECPoint pubkey)
...
-            NeoAccountState state_account = storage_account.GetInteroperable<NeoAccountState>();
+            NeoAccountState state_account = engine.Snapshot.Storages.GetAndChange(CreateStorageKey(Prefix_Account).Add(account))?.GetInteroperable<NeoAccountState>();
+            if (state_account is null) return false;
+            CandidateState validator_new = null;
+            if (voteTo != null)
+            {
+                validator_new = engine.Snapshot.Storages.GetAndChange(CreateStorageKey(Prefix_Candidate).Add(voteTo))?.GetInteroperable<CandidateState>();
+                if (validator_new is null) return false;
+                if (!validator_new.Registered) return false;
+            }
```
Comment (cloud8little) [inline src/neo/SmartContract/Native/Tokens/NeoToken.cs:152]: why not throw an exception?

Comment (erikzhang) [reply src/neo/SmartContract/Native/Tokens/NeoToken.cs:152]: Returning `false` can give the contract a chance to recover from the exception. Throwing an exception will directly terminate the virtual machine. we should try not to throw exceptions if the consistency can be guaranteed.

**Comment by cloud8little** [comment]
Retest passed. 
1. vote when pubkey register
```
neo> vote Nb4z5KTAqgavWKZ7RemRS73RGkY8Qc68mm 0358c84fee21b68e568b1a43d1cf6bddf74cce08ad5f3b301f3ad6d3ad75a9a6dc
Invoking script with: '0c210358c84fee21b68e568b1a43d1cf6bddf74cce08ad5f3b301f3ad6d3ad75a9a6dc0c14a63e5f5deb760927fed26f64e40cabff2309329312c00c04766f74650c1425059ecb4878d3a875f91c51ceded330d4575fde41627d5b52'
VM State: HALT
Gas Consumed: 5.0100793
Evaluation Stack: [{"type":"Boolean","value":true}]

relay tx(no|yes): yes
Signed and relayed transaction with hash=0x590578ebe8242810f28552a02d339ec9b5e54b3a553ed602df80fe49b9ea2be4
```
2. vote before register, then send neo successfully to other accounts.
```
neo> vote Nb4z5KTAqgavWKZ7RemRS73RGkY8Qc68mm 027a4c687d168044dfb37ad32a51830f0c86b068084914f2373702393b1923ae5e
Invoking script with: '0c21027a4c687d168044dfb37ad32a51830f0c86b068084914f2373702393b1923ae5e0c14a63e5f5deb760927fed26f64e40cabff2309329312c00c04766f74650c1425059ecb4878d3a875f91c51ceded330d4575fde41627d5b52'
VM State: HALT
Gas Consumed: 5.0100793
Evaluation Stack: [{"type":"Boolean","value":false}]

relay tx(no|yes): yes
Signed and relayed transaction with hash=0xdc4c34f6f537c5b11389425584269288f97eb04f2d69f52372833944b60ca85d
```

**Comment by superboyiii** [comment]
Looks good now.

**Comment by superboyiii** [comment]
@erikzhang Shall we merge?

**Comment by superboyiii** [comment]
@shargon Need your merge.

**Pull request merged**